Firestore/core/include/firebase/firestore/document_reference.h

```diff
@@ -0,0 +1,362 @@
+   * this DocumentReference.
+   */
+  virtual Future<DocumentSnapshot> Get() const;
+
+  /**
+   * @brief Writes to the document referred to by this DocumentReference.
+   *
+   * If the document does not yet exist, it will be created. If you pass
+   * SetOptions, the provided data can be merged into an existing document.
+   *
+   * @param[in] data A map of the fields and values for the document.
+   * @param[in] options An object to configure the set behavior.
+   *
+   * @return A Future that will be resolved when the write finishes.
+   */
+  virtual Future<void> Set(const MapFieldValue& data,
+                           const SetOptions& options = SetOptions());
+
+  /**
+   * @brief Updates fields in the document referred to by this
+   * DocumentReference.
+   *
+   * If no document exists yet, the update will fail.
+   *
+   * @param[in] data A map of field / value pairs to update. Fields can contain
+   * dots to reference nested fields within the document.
+   *
+   * @return A Future that will be resolved when the write finishes.
+   */
+  virtual Future<void> Update(const MapFieldValue& data);
+
+  /**
+   * @brief Removes the document referred to by this DocumentReference.
+   *
+   * @return A Task that will be resolved when the delete completes.
+   */
+  virtual Future<void> Delete();
+
+  /**
+   * @brief Starts listening to the document referenced by this
+   * DocumentReference.
+   *
+   * @param[in] listener The event listener that will be called with the
+   * snapshots, which must remain in memory until you remove the listener from
+   * this DocumentReference. (Ownership is not transferred; you are responsible
+   * for making sure that listener is valid as long as this DocumentReference is
+   * valid and the listener is registered.)
+   *
+   * @return A registration object that can be used to remove the listener.
+   */
+  virtual ListenerRegistration AddSnapshotListener(
+      EventListener<DocumentSnapshot>* listener);
+
+  /**
+   * @brief Starts listening to the document referenced by this
+   * DocumentReference.
+   *
+   * @param[in] options The options to use for this listen.
+   * @param[in] listener The event listener that will be called with the
+   * snapshots, which must remain in memory until you remove the listener from
+   * this DocumentReference. (Ownership is not transferred; you are responsible
+   * for making sure that listener is valid as long as this DocumentReference is
+   * valid and the listener is registered.)
+   *
+   * @return A registration object that can be used to remove the listener.
+   */
+  virtual ListenerRegistration AddSnapshotListener(
+      const DocumentListenOptions& options,
+      EventListener<DocumentSnapshot>* listener);
+
+#if defined(FIREBASE_USE_STD_FUNCTION) || defined(DOXYGEN)
+  /**
+   * @brief Starts listening to the document referenced by this
+   * DocumentReference.
+   *
+   * @param[in] callback function or lambda to call.
```

Comment:
Consider mentioning that (I assume) only one of the pointers passed to the callback is non-null.

Reply:
Done